Studio: Make the import process more verbose

bcotrim opened the pull request:
## Related issues

- Fixes STU-746

## Proposed Changes

- Added detailed progress tracking during import with file counts and type-specific messages
- Added detailed progress tracking during site creation, with messages from Playground CLI.
- Enhanced extraction progress to show "Extracting backup… (XX%)"
- Added database import details showing "Importing database… (XX%)"
- Implemented WordPress content breakdown by type: "Importing plugins… (XX%)", "Importing themes… (XX%)", "Importing media uploads… (XX%)"
- Added new progress events: BACKUP_EXTRACT_FILE_START, IMPORT_DATABASE_PROGRESS, IMPORT_WP_CONTENT_PROGRESS
- Updated backup handlers (ZIP, tar.gz) to emit granular progress data
- Fixed unit test expectations for tar.gz handler

## Testing Instructions

- Create a test site with multiple plugins, themes, and content (use FakerPress for posts with images)
- Verify the create site window shows detailed progress messages
- Export the site as a backup
- Import into a new site and observe detailed progress messages
- Test with different backup types (ZIP, tar.gz, .wpress)
- Verify progress shows file counts and content type information
- For slower testing to observe messages, temporarily add delays in handlers (remove before merge)

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

bcotrim committed:
make the import process more verbose

bcotrim committed:
update unit tests

bcotrim committed:
update handlers from wpress and sql

gcsecsey commented on src/lib/import-export/import/importers/importer.ts:
```diff
@@ -194,32 +208,73 @@ abstract class BaseBackupImporter extends BaseImporter {
+		let processedItems = 0;
+		const totalItems = this.backup.wpContentFiles.length;
+
+		for ( const [ type, files ] of Object.entries( filesByType ) ) {
+			for ( const file of files ) {
```
I think tracking the progress this way makes sense, because we categorized the files into these types in the first place.

gcsecsey reviewed (commented):
This looks good and IMO the added progress steps make the import process much more "interactive" from a UX perspective, it's great to see what's happening under the hood.

To test the changes, I configured a site on playground.wordpress.net with fake data and plugins, and exported it both as .wpress and .zip.

When importing the .wpress I ran into an SQL parse error at the end of the import process. I could recreate this error on trunk, so I don't think this is related to these changes:

https://github.com/user-attachments/assets/2dca9f04-e211-4305-b8a4-f77d0cd09f64

[error-log.log](https://github.com/user-attachments/files/22253464/error-log.log)

Importing the .zip file worked well without errors:

https://github.com/user-attachments/assets/b1d8398e-aab5-4230-9cc0-90cbaa99b64a

bcotrim committed:
Address PR review comments: remove double file read, use proper event emission, improve path categorization

bcotrim committed:
Address additional PR comments: restore totalFiles for consistency, extract progress helper, optimize typeLabels constant

bcotrim committed:
Fix progress display to show file counts starting from 0

bcotrim committed:
Emit progress event after file copy to prevent UI freeze appearance

bcotrim committed:
add messages to create site

bcotrim committed:
fix lint

bcotrim committed:
cleanup comments

epeicher reviewed (commented):
Thanks @bcotrim! This now looks much better, I have tested with a `.zip` file and I have not found any issues, I can see the total file count and the progress is much better now

### Importing a zip

https://github.com/user-attachments/assets/37356740-93b6-4b92-b5b2-0d55664fc704

But I have found a small pause in the progress when importing a `wpress` file in this branch that I could not replicate in `trunk`. Please find videos below

### Importing a wpress in this branch

https://github.com/user-attachments/assets/d0f316e9-16a8-4031-8d6e-580550e2c94e


### Importing a wpress in trunk

https://github.com/user-attachments/assets/8ebb8dcd-e1c4-47c2-98b0-3d35b4b3524c

Please ping me if you need the `wpress` file

wojtekn commented:
@bcotrim, this will be a nice improvement!

It will prepare Studio for another change - at some point, we can consider redesigning how Studio displays progress, to:
- display all steps on the screen upfront
- preserve each completed step with something that marks it as complete

That way, Studio would set expectations, so users would see all the steps that will be performed to complete the import. (CC @Marinatsu )

One note so far regarding UI - would it be clearer to use a percentage in all those places? As a user, I wouldn't be interested in the file count, but rather in the overall process progress:

> Extracting backup… (19%)
> Importing database… (67%)
> Importing plugins… (42%)
> Importing themes… (67%)
> Importing media uploads… (30%)

We could also consider changing labels for the create site steps to make them more meaningful. For now, ale steps from "Getting everything ready…" to "Nearly finished" add some impression of progress but they don't mean anything.

bcotrim committed:
change to percentages

bcotrim committed:
add start site messages

bcotrim commented:
@wojtekn @epeicher @gcsecsey Thanks for the reviews!

I addressed the comments and issues.
Added new behavior:
- (XX%) instead of number of files
- Detailed progress messages for site creation process

Can you take another look, please?

epeicher commented on src/lib/wordpress-provider/playground-cli/playground-server-process-child.ts:
```diff
@@ -48,9 +48,24 @@ const originalConsoleLog = console.log;
 const originalConsoleError = console.error;
 const originalConsoleWarn = console.warn;
 
+function formatMessageForUI( message: string ): string | null {
```
Nit: Looking at the return type, can this function return `null`?

epeicher reviewed (commented):
Thanks @bcotrim for continuing to improve this! I have tested it and it works great, I like the percentages, and when creating a site from an import, the progress is much nicer now. 

https://github.com/user-attachments/assets/5c080c48-806e-40b7-8571-56e43ea23803

With regards to importing a backup to an existing site, I have found a couple of details that could be improved as part of this, or as follow-ups:

1. When importing a `.wpress` backup, I see a small pause in the import (you can see in the video around the second 3) that I don't see in trunk 

https://github.com/user-attachments/assets/78057de9-9201-4553-a708-d462002ca34f

2. The progress bar goes backwards during the different states, as a user I don't like that, I would prefer an indeterminate progress bar or ideally a progress bar that only goes forward. Although, this could be a matter of preference, so happy to hear others thoughts.

bcotrim committed:
pr feedback

bcotrim committed:
fix unit tests

epeicher commented on src/lib/wordpress-provider/playground-cli/playground-server-process.ts:
```diff
@@ -62,14 +64,15 @@ export class PlaygroundServerProcess implements WordPressServerProcess {
-								siteId,
+					if ( this.siteId ) {
+						try {
+							const mainWindow = await getMainWindow();
```
👍

epeicher reviewed (approved):
Thanks @bcotrim for addressing the comments! I’ve given it another spin, and it works great. I don’t see any pauses in the progress, and I can see the progress bar is indeterminate, but the percentage gives a nice sense of progress. This LGTM! :shipit:

### Creating a  new site by importing a wpress file

https://github.com/user-attachments/assets/9e3be898-40c9-4629-aded-9fa17e67f31f

### Importing a wpress file into an existing site

https://github.com/user-attachments/assets/b595517c-d7ca-411f-9945-09a7b0dd3417

I have also tested with a jetpack backup and it works great too 🙌
